gotham/src/test/mod.rs

```diff
@@ -189,45 +168,60 @@ where
     ///
     /// If the future came from a different instance of `TestServer`, the event loop will run until
     /// the timeout is triggered.
-    fn run_request<F>(&self, f: F) -> Result<F::Item, TestRequestError>
+    fn run_request<F>(&mut self, f: F) -> Result<F::Item>
     where
-        F: Future<Error = hyper::Error>,
+        F: Future + Send + 'static,
+        F::Error: failure::Fail + Sized,
+        F::Item: Send,
     {
         let timeout_duration = Duration::from_secs(self.data.timeout);
-        let timeout = Timeout::new(timeout_duration, &self.data.core.borrow().handle())
-            .map_err(|e| TestRequestError::IoError(e))?;
-
-        let run_result = {
-            let mut core = self.data.core.borrow_mut();
-            core.run(f.select2(timeout))
-        };
+        let timeout = Delay::new(timeout_duration);
 
-        match run_result {
-            Ok(future::Either::A((item, _))) => Ok(item),
-            Ok(future::Either::B(_)) => Err(TestRequestError::TimedOut),
-            Err(future::Either::A((e, _))) => Err(TestRequestError::HyperError(e)),
-            Err(future::Either::B((e, _))) => Err(TestRequestError::IoError(e)),
+        match self.run_future(f.select2(timeout).map_err(|either| {
```

Comment:
The readability here is a bit... crazy. Might be a nice idea to store the future separately above before the `match`.